SMQ-3234 - Add notifications service

WashingtonKK opened the pull request:
<!-- Copyright (c) Abstract Machines
SPDX-License-Identifier: Apache-2.0 -->

<!--

Pull request title should be `SMQ-XXX - description` or `NOISSUE - description` where XXX is ID of the issue that this PR relate to.
Please review the [CONTRIBUTING.md](https://github.com/absmach/supermq/blob/main/CONTRIBUTING.md) file for detailed contributing guidelines.

For Work In Progress Pull Requests, please use the Draft PR feature, see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

For a timely review/response, please avoid force-pushing additional commits if your PR already received reviews or comments.

- Provide tests for your changes.
- Use descriptive commit messages.
- Comment your code where appropriate.
- Squash your commits
- Update any related documentation.
-->

# What type of PR is this?

<!--This represents the type of PR you are submitting.

For example:
This is a bug fix because it fixes the following issue: #1234
This is a feature because it adds the following functionality: ...
This is a refactor because it changes the following functionality: ...
This is a documentation update because it updates the following documentation: ...
This is a dependency update because it updates the following dependencies: ...
This is an optimization because it improves the following functionality: ...
-->

## What does this do?

<!--
Please provide a brief description of what this PR is intended to do.
Include List any changes that modify/break current functionality.
-->

## Which issue(s) does this PR fix/relate to?

<!--
For pull requests that relate or close an issue, please include them below.  We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "Resolves #1234" would connect the current pull request to issue 1234.  And when we merge the pull request, Github will automatically close the issue.
-->

- Related Issue #
- Resolves #

## Have you included tests for your changes?

<!--If you have not included tests, please explain why.
For example:
Yes, I have included tests for my changes.
No, I have not included tests because I do not know how to.
-->

## Did you document any new/modified feature?

<!--If you have not included documentation, please explain why.
For example:
Yes, I have updated the documentation for the new feature.
No, I have not updated the documentation because I do not know how to.
-->

### Notes

<!--Please provide any additional information you feel is important.-->

WashingtonKK committed:
SMQ-3234 - Add notifications service

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

SammyOina commented on notifications/emailer/emailer.go:
```diff
@@ -0,0 +1,54 @@
+	"github.com/absmach/supermq/notifications"
+)
+
+var _ notifications.Emailer = (*emailer)(nil)
```
you should have a notifier interface and implmentations  for different notifiers (SMS, email, slack ...)

SammyOina commented on notifications/emailer/emailer.go:
```diff
@@ -0,0 +1,54 @@
+	}, nil
+}
+
+func (e *emailer) SendInvitationSentEmail(to []string, inviteeName, domainName, inviterName string) error {
```
One send method is enough and message strcture can be defined  with events especially subject and headers need to be dynamic to avoid grouping messages from prism and smq in yahoo

SammyOina commented on notifications/emailer/emailer.go:
```diff
@@ -0,0 +1,54 @@
+	subject := fmt.Sprintf("You've been invited to join %s", domainName)
+	header := "Domain Invitation"
+	content := fmt.Sprintf("%s has invited you to join the domain %s. Please log in to accept or reject this invitation.", inviterName, domainName)
+	footer := "SuperMQ Team"
```
needs to be dynamic

dborovcanin reviewed (changes requested):
@WashingtonKK Check our old notifier approach:
https://github.com/absmach/supermq-contrib/tree/main/consumers/notifiers.

Notification service is **consumer**, i.e. we subscribe to events. What we need is:
1. actual Notification sender - think something like an interface like:
```go
type Notifier interface {
    Notify(Notification) error
}
```
So, notification struct is 👍🏽.
2. Notification service contains a collection of notifiers, think of it like email, Slack, SMS, any other notifier. Notification service is used to subscribe to provided events, convert them to Notification, and route Notification to the registered notifiers.

Since notifications are not decoupled from other services (nobody should be importing them), you can import users or any other SMQ package directly.

WashingtonKK committed:
refactor notifications service

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
refactor notifications service

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
update mocks

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
add consumers tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

SammyOina commented on users/emailer/emailer.go:
```diff
@@ -48,3 +62,15 @@ func (e *emailer) SendVerification(to []string, user, verificationToken string)
+	subject := fmt.Sprintf("%s accepted your invitation to %s", inviteeName, domainName)
+	content := fmt.Sprintf("%s has accepted your invitation to join %s as %s.", inviteeName, domainName, roleName)
+	return e.invitationAcceptedAgent.Send(to, "", subject, "", inviterName, content, "SuperMQ Team")
+}
```
we just need one send/ notify method

SammyOina commented on users/events/consumer.go:
```diff
@@ -0,0 +1,233 @@
+	case invitationSend:
+		return handleInvitationSent(ctx, data, h.emailer, h.userRepo, h.logger)
+	case invitationAccept:
+		return handleInvitationAccepted(ctx, data, h.emailer, h.userRepo, h.logger)
```
there should be more events here

SammyOina replied:
there should be more events here

SammyOina replied:
there should be more events here

SammyOina commented on users/emailer.go:
emailer should implement notifier interface rather than having its own interface

dborovcanin replied:
This has not been addressed.

WashingtonKK committed:
refactor: consolidate notifications into users service with Notifier abstraction

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
refactor: use single notifier interface for all abstraction

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

dborovcanin commented on users/notifier.go:
```diff
@@ -0,0 +1,38 @@
+)
+
+// NotificationData contains the data needed to send a notification.
+type NotificationData struct {
```
Rename to Notification.

dborovcanin commented on users/service.go:
```diff
@@ -41,17 +41,17 @@ type service struct {
 
 // NewService returns a new Users service implementation.
-func NewService(token grpcTokenV1.TokenServiceClient, urepo Repository, policyService policies.Service, emailer Emailer, hasher Hasher, idp supermq.IDProvider) Service {
+func NewService(token grpcTokenV1.TokenServiceClient, urepo Repository, policyService policies.Service, notifier Notifier, hasher Hasher, idp supermq.IDProvider) Service {
```
Ideally, there would be a slice of notifiers for different types of notifications. It's OK to keep it a single Notifier since we are talking about emailer only.

dborovcanin commented on users/emailer/notifier.go:
```diff
@@ -0,0 +1,72 @@
+	}
+}
+
+func (n *emailNotifier) notifyPasswordReset(data users.NotificationData) error {
```
Always propagate context from the caller (Notify).

dborovcanin commented on users/events/consumer.go:
```diff
@@ -0,0 +1,255 @@
+	data, err := event.Encode()
+	if err != nil {
+		h.logger.Error("failed to encode event", slog.Any("error", err))
+		return nil
```
Why don't we return an error?

dborovcanin commented on users/emailer/notifier.go:
```diff
@@ -0,0 +1,72 @@
+	inviteeName := data.Metadata["invitee_name"]
+	inviterName := data.Metadata["inviter_name"]
+	domainName := data.Metadata["domain_name"]
+	roleName := data.Metadata["role_name"]
```
We are missing a lot of validation here. Also, DRY, decode once, and validate depending on the kind of the email you want to send. `notifyInvitationSent` and `notifyInvitationAccepted` are the same. What about rejected?

dborovcanin commented on users/events/consumer.go:
```diff
@@ -0,0 +1,255 @@
+
+	// Normalize names for display
+	inviteeName := invitee.FirstName + " " + invitee.LastName
+	if inviteeName == " " || inviteeName == "" {
```
`inviteeName == ""` will never be the case because of the previous line.

dborovcanin commented on users/notifier.go:
```diff
@@ -0,0 +1,38 @@
+	Recipients []string
+
+	// Metadata contains additional data specific to the notification type.
+	Metadata map[string]string
```
While very generic, this way makes handling notification more cryptic and complex then it needs to be. I.e. I'd prefer even `Notify(ctx context.Context, notification any) error` over extracting it from the map and expecting keys to be there - we'd have at least type safety once we cast `any` and instead of `NotificationData.Type` use `switch notification.(type)`. Check two other (unlikely) options:
1. Making Notification an interface (notifications holds data, so it's probably a dead-end)
2. Using generics (probably will end up mostly the same code as `Notify(ctx context.Context, notification any) error`)

WashingtonKK committed:
refactor: implement type-safe notification system with rejection support

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
refactor invitation notification

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

SammyOina commented on users/emailer/emailer.go:
```diff
@@ -4,23 +4,39 @@
+	verifyAgent             *email.Agent
+	invitationAgent         *email.Agent
+	invitationAcceptedAgent *email.Agent
+	invitationRejectedAgent *email.Agent
```
we do not need multiple email agents

WashingtonKK replied:
https://github.com/absmach/supermq/issues/3257

SammyOina commented on users/emailer/emailer.go:
```diff
@@ -31,20 +47,142 @@ func New(resetURL, verificationURL string, resetConfig, verifyConfig *email.Conf
-func (e *emailer) SendVerification(to []string, user, verificationToken string) error {
-	url := fmt.Sprintf("%s?token=%s", e.verificationURL, verificationToken)
-	return e.verifyAgent.Send(to, "", "Email Verification", "", user, url, "")
+func (e *emailer) sendEmailVerification(_ context.Context, notif *users.EmailVerificationNotification) error {
```
don't pass context if not used

SammyOina replied:
same in other places

SammyOina commented on users/events/consumer.go:
```diff
@@ -0,0 +1,343 @@
+}
+
+func handleInvitationRejected(ctx context.Context, data map[string]any, notifier users.Notifier, userRepo users.Repository, logger *slog.Logger) error {
+	inviteeUserID, _ := data["invitee_user_id"].(string)
```
will panic if assetion is not valid, do a check

WashingtonKK committed:
fix tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
refactor

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

SammyOina commented on users/events/consumer.go:
```diff
@@ -0,0 +1,402 @@
+	}
+}
+
+func handleInvitationSent(ctx context.Context, data map[string]any, notifier users.Notifier, userRepo users.Repository, logger *slog.Logger) error {
```
this functions seems to only ever return nil and never any error

SammyOina commented on users/events/consumer.go:
```diff
@@ -0,0 +1,402 @@
+	return nil
+}
+
+func handleInvitationAccepted(ctx context.Context, data map[string]any, notifier users.Notifier, userRepo users.Repository, logger *slog.Logger) error {
```
same here

SammyOina commented on users/events/consumer.go:
```diff
@@ -0,0 +1,402 @@
+	return nil
+}
+
+func handleInvitationRejected(ctx context.Context, data map[string]any, notifier users.Notifier, userRepo users.Repository, logger *slog.Logger) error {
```
same here, you type assertion and if they fail you just continue with the invalid data type, meaning there is likely to be a runtime panic

SammyOina commented on internal/email/email.go:
```diff
@@ -106,5 +107,7 @@ func (a *Agent) Send(to []string, from, subject, header, user, content, footer s
 		return errors.Wrap(errSendMail, err)
 	}
 
+	fmt.Println("Email Sent")
```
clean up the pr

WashingtonKK committed:
return errors on events package

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>
Signed-off-by: dusan <borovcanindusan1@gmail.com>

WashingtonKK committed:
lint

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>
Signed-off-by: dusan <borovcanindusan1@gmail.com>

SammyOina commented on users/events/consumer.go:
```diff
@@ -0,0 +1,391 @@
+
+	operation, ok := data["operation"].(string)
+	if !ok {
+		return nil
```
should not be nil if type assertion failed

SammyOina commented on users/events/consumer.go:
```diff
@@ -0,0 +1,391 @@
+			slog.String("user_id", inviteeUserID),
+			slog.Any("error", err),
+		)
+		return nil
```
why return nil here

SammyOina commented on users/events/consumer.go:
```diff
@@ -0,0 +1,391 @@
+			slog.String("user_id", inviteeUserID),
+			slog.Any("error", err),
+		)
+		return nil
```
a number of lines where we have an error but return nil

SammyOina commented on users/events/consumer.go:
```diff
@@ -0,0 +1,391 @@
+		return nil
+	}
+
+	logger.Info("invitation rejected notification sent",
```
I see a lot of logging, we can do this in middleware

WashingtonKK committed:
address comments

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>
Signed-off-by: dusan <borovcanindusan1@gmail.com>

WashingtonKK committed:
remove excessive logs

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>
Signed-off-by: dusan <borovcanindusan1@gmail.com>

arvindh123 commented on users/notifier.go:
```diff
@@ -0,0 +1,48 @@
+type Notifier interface {
+	// Notify sends a notification based on the notification type.
+	Notify(ctx context.Context, notification any) error
+}
```
This can be emailer, because it sends email not notifications

arvindh123 committed:
remove invitations event handler and invitations emailer from users service

Signed-off-by: Arvindh <arvindh91@gmail.com>
Signed-off-by: dusan <borovcanindusan1@gmail.com>

arvindh123 committed:
remove invitations event handler and invitations emailer from users service

Signed-off-by: Arvindh <arvindh91@gmail.com>
Signed-off-by: dusan <borovcanindusan1@gmail.com>

arvindh123 committed:
feat(users): add gRPC retrieval of users by IDs

Signed-off-by: Arvindh <arvindh91@gmail.com>
Signed-off-by: dusan <borovcanindusan1@gmail.com>

arvindh123 committed:
Add users service grpc endpoint env varaible

Signed-off-by: Arvindh <arvindh91@gmail.com>
Signed-off-by: dusan <borovcanindusan1@gmail.com>

dborovcanin committed:
Add notifications service for invitation events

Add new notifications service that listens to domain invitation events
(send, accept, reject) and sends email notifications to users. The service
uses gRPC to fetch user information from the users service and sends
beautifully styled HTML emails with Magistrala branding (#083662).

Signed-off-by: dusan <borovcanindusan1@gmail.com>

dborovcanin committed:
Optimize user fetching with batch gRPC request

Optimize the emailer to fetch both inviter and invitee users in a single
gRPC call instead of making two separate requests. This reduces network
overhead and improves performance.

Signed-off-by: dusan <borovcanindusan1@gmail.com>

dborovcanin committed:
Add observability middleware and fix shutdown

Signed-off-by: dusan <borovcanindusan1@gmail.com>

WashingtonKK committed:
Fix linting issues in notifications package

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>
Signed-off-by: dusan <borovcanindusan1@gmail.com>

WashingtonKK committed:
Add docker config for notifications

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>
Signed-off-by: dusan <borovcanindusan1@gmail.com>

dborovcanin committed:
Add mocks for Notifier service

Signed-off-by: dusan <borovcanindusan1@gmail.com>

dborovcanin committed:
Fix tests

Signed-off-by: dusan <borovcanindusan1@gmail.com>

dborovcanin committed:
Unify handling of the invitation events

Signed-off-by: dusan <borovcanindusan1@gmail.com>

arvindh123 commented on notifications/events/consumer.go:
```diff
@@ -0,0 +1,112 @@
+}
+
+func parseNotificationFromEvent(event events.Event, errorContext string) (notifications.Notification, error) {
+	data, err := event.Encode()
+	if err != nil {
+		slog.Error(fmt.Sprintf("failed to encode %s event", errorContext), "error", err)
+		return notifications.Notification{}, err
+	}
+
+	invitedBy, ok := data["invited_by"].(string)
+	if !ok || invitedBy == "" {
+		slog.Error(fmt.Sprintf("missing or invalid invited_by in %s event", errorContext))
+		return notifications.Notification{}, fmt.Errorf("missing or invalid invited_by")
+	}
+
+	inviteeUserID, ok := data["invitee_user_id"].(string)
+	if !ok || inviteeUserID == "" {
+		slog.Error(fmt.Sprintf("missing or invalid invitee_user_id in %s event", errorContext))
+		return notifications.Notification{}, fmt.Errorf("missing or invalid invitee_user_id")
+	}
+
+	domainID, ok := data["domain_id"].(string)
+	if !ok || domainID == "" {
+		slog.Error(fmt.Sprintf("missing or invalid domain_id in %s event", errorContext))
+		return notifications.Notification{}, fmt.Errorf("missing or invalid domain_id")
+	}
+
+	roleID, _ := data["role_id"].(string)
+	domainName, _ := data["domain_name"].(string)
+	roleName, _ := data["role_name"].(string)
+
+	return notifications.Notification{
+		InviterID:  invitedBy,
+		InviteeID:  inviteeUserID,
+		DomainID:   domainID,
+		DomainName: domainName,
+		RoleID:     roleID,
+		RoleName:   roleName,
+	}, nil
+}
```
Can we move this parser to common package `pkg/domains/events/consumer/decode.go`

dborovcanin replied:
For the time being, let's keep it here. If we introduce more kinds of notification, it will make sense to move it to a shared decoder.

dborovcanin committed:
Fix tests

Signed-off-by: dusan <borovcanindusan1@gmail.com>

dborovcanin committed:
Make decoding safer

Signed-off-by: dusan <borovcanindusan1@gmail.com>

dborovcanin committed:
Fix domains invitation handling

Signed-off-by: dusan <borovcanindusan1@gmail.com>

dborovcanin committed:
Improve domains and notifictaions code

Signed-off-by: dusan <borovcanindusan1@gmail.com>

SammyOina reviewed (changes requested):
<img width="2296" height="632" alt="Image" src="https://github.com/user-attachments/assets/15cc094a-f2d8-4ebd-a037-f99add2e2272" />

subject on verification will still result in grouping on yahoo

dborovcanin committed:
Fix method name and formatting

Signed-off-by: dusan <borovcanindusan1@gmail.com>

arvindh123 commented on notifications/notifier.go:
```diff
@@ -0,0 +1,37 @@
+)
+
+// Notification contains the data needed to send a notification.
+type Notification struct {
+	Type       NotificationType
+	InviterID  string
+	InviteeID  string
+	DomainID   string
+	DomainName string
+	RoleID     string
+	RoleName   string
+}
```
Is this specific to invitations ?

arvindh123 commented on notifications/events/consumer.go:
```diff
@@ -0,0 +1,134 @@
+
+const (
+	// Stream names.
+	sendInvitationStream   = "events.supermq.invitation.send"
+	acceptInvitationStream = "events.supermq.invitation.accept"
+	rejectInvitationStream = "events.supermq.invitation.reject"
```
@dborovcanin
Just saying my thoughts 

Having stream topic as constant make notification service works only for invitations 
We need to make this configurable via file 

The file should have event topic,event payload parser, notification condition check , then emailer or slacker or webhooker

dborovcanin committed:
Fix domain tests

Signed-off-by: dusan <borovcanindusan1@gmail.com>

dborovcanin committed:
Add domain alternative name

Signed-off-by: dusan <borovcanindusan1@gmail.com>

dborovcanin committed:
Fix tests

Signed-off-by: dusan <borovcanindusan1@gmail.com>
